tools/flaky-test-reporter/slack_notification.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// slack_notification.go sends notifications to slack channels
+
+package main
+
+import (
+	"io/ioutil"
+	"strings"
+	"time"
+	"log"
+	"fmt"
+	"encoding/json"
+
+	"net/http"
+	"net/url"
+
+	"github.com/knative/test-infra/shared/testgrid"
+)
+
+const (
+	slackChatPostMessageURL = "https://slack.com/api/chat.postMessage"
+	// default filter for testgrid link
+	testgridFilter          = "exclude-non-failed-tests=20"
+)
+
+var (
+	slackUsername = "Knative bot"
+	slackTokenStr string
+	slackIconEmoji *string
+)
+
+// jobNameTestgridURLMap contains harded coded mapping of job name: Testgrid tab URL relative to base URL
+var jobNameTestgridURLMap = map[string]string {
+	"ci-knative-serving-continuous": "knative-serving#continuous",
+}
+
+// slackChannel contains channel logical name and Slack identity
+type slackChannel struct {
+	name, identity string
+}
+
+// slackChannelsMap defines mapping of repo: slack channels
+var slackChannelsMap = map[string][]slackChannel{
+	"serving" : []slackChannel{{"test", "CA1DTGZ2N"}},
+}
+
+// authenticateSlack reads token file and stores it for later authentication
+func authenticateSlack(slackTokenPath string) error {
+	b, err := ioutil.ReadFile(slackTokenPath)
+	if err != nil {
+		return err
+	}
+	slackTokenStr = strings.TrimSpace(string(b))
+	return nil
+}
+
+// postMessage does http post
+func postMessage(url string, uv *url.Values) error {
```

Comment:
these messages should be on the slack client and not independent functions

Reply:
done